ADR/00xx-component-references.md

```diff
@@ -0,0 +1,155 @@
+  * This is a many-to-one dependency. Many component images depend on one common parent image.
+* A user team has an **OLM operator**. When they propose an update to one of their operands with new code, they want to see if that’s going to break their operator. In order to be fully tested, a single rebuilt image needs to be included as a reference in a bundle image in order to be tested as a whole unit when deployed via OLM, i.e. for integration tests ([RHTAP-992](https://issues.redhat.com/browse/RHTAP-992)). A real example of this is in [gatekeeper](https://github.com/gatekeeper/gatekeeper-operator) where the operator repo contains both the [controller code](https://github.com/gatekeeper/gatekeeper-operator/blob/main/controllers/gatekeeper_controller.go) and the [bundle metadata](https://github.com/gatekeeper/gatekeeper-operator/blob/main/config/manifests/bases/gatekeeper-operator.clusterserviceversion.yaml), which need to be built into separate images([1](https://github.com/gatekeeper/gatekeeper-operator/blob/main/Dockerfile) and [2](https://github.com/gatekeeper/gatekeeper-operator/blob/main/bundle.Dockerfile)), separate Components in AppStudio.
+  * In this case, the bundle image must be rebuilt for the operand image update to be tested at all (assuming OLM is the strategy for deploying the operator).
+  * This is a one-to-many dependency. One operator bundle depends on many operand images.
+* A user team just has **two components that depend on each other in a functional way**. When they propose an update to one component, they want to submit a corresponding change to the second component, and have those tested together before merging both.
+  * In this case, on any given day, component B’s PRs may or may not have dependencies on PRs from component A. Not all changes are linked.
+  * This is a many-to-many dependency. The user may have lots of components that depend on lots of other components.
+
+In theory **any combination of the cases above** could be present in an app. They could have one common parent image, many operand images that depend on that, and one or more operator bundles which depend on those.
+
+Today, users work around how complicated it is to manage digests themselves by instead using *floating tags*, which have the benefit of being easy to use - no need to update - but have the problem of being unclear. It's not exactly clear what you're building against if you refer to it by tag. Potentially insecure. We want to make it easy for users to do better.
+
+## Decision
+
+### Interface changes
+
+**Declared Reference Embeddings**: Introduce a new CustomResource called `ComponentEmbedding` that lets one Component declare that another Component **embeds a reference** to it.
+
+* A `ComponentEmbedding` spec contains two fields: a embedded component, and a embedding component
```

Comment:
Can we explicitly say that both those fields should be a list to simplify configuration?

We need a one-to-many for parent images, and many-to-one for bundles.

Reply:
Oh - I guess I thought that they would be scalar references, not lists. I.e., if you have Component A that embeds a reference to Component B that embeds a reference to Component C, you would have 3 Components and 2 ComponentEmbeddings. One ComponentEmbedding for A->B and another ComponentEmbedding for B->C.

Reply:
Ack. Just to clarify, the proposal as it is written in light of your common image (Component A) used by 3 other components (B, C, D), would look like 4 Components and 3 ComponentEmbeddings, one for B->A, one for C->A, and one for D->A.

Reply:
Correct. In my opinion the list provides a better view of what's happening.

However I could see a UI building a graph to show the dependencies relations to the user, and buttons to add/delete relationships between components. At that point how the data is stored would not be very relevant anymore, especially if the user is not expected to interact with CRs directly to update them. It might be easier to deal with while troubleshooting though?

Reply:
Yeah - I think both models can and will work. I hesitate to dictate that one should be the true way here in the ADR. If we say that ComponentEmbedding is owned by 'build-service` or `application-service` (and should be defined in the redhat-appstudio/application-api repo), then perhaps we should defer to the implementing team to decide which model to implement.

Reply:
Agreed. We can close this thread.

Reply:
@Roming22 after consulting with the HAS team, I ended up making a change similar to but different from what you proposed in 40581df6bb0e046989b90eb43236c8716e1253e3.